File: backend/services/vnpayService.js
const crypto = require('crypto');
const querystring = require('querystring');
const moment = require('moment');

/**
 * VNPay Payment Gateway Service
 * 
 * Handles VNPay Sandbox integration for payment processing
 * API Documentation: https://sandbox.vnpayment.vn/apis/docs/
 * 
 * Test Credentials (Sandbox):
 * - Card Number: 9704198526191432198
 * - Card Holder: NGUYEN VAN A
 * - Expiry Date: 07/15
 * - OTP: 123456
 */
class VNPayService {
  constructor() {
    // VNPay Sandbox URL
    this.vnpUrl = 'https://sandbox.vnpayment.vn/paymentv2/vpcpay.html';
    this.vnpApiUrl = 'https://sandbox.vnpayment.vn/merchant_webapi/api/transaction';
    
    // Credentials from .env
    this.tmnCode = process.env.VNP_TMN_CODE;
    this.hashSecret = process.env.VNP_HASH_SECRET;
    this.returnUrl = process.env.VNP_RETURN_URL;
    
    // Validate configuration
    if (!this.tmnCode || !this.hashSecret || !this.returnUrl) {
      console.warn('⚠️  VNPay configuration missing in .env file');
    }
  }

  /**
   * Sort object keys alphabetically (required by VNPay)
   * @param {Object} obj - Object to sort
   * @returns {Object} Sorted object
   */
  sortObject(obj) {
    const sorted = {};
    const keys = Object.keys(obj).sort();
    
    keys.forEach(key => {
      sorted[key] = obj[key];
    });
    
    return sorted;
  }

  /**
   * Create VNPay payment URL
   * 
   * @param {string} orderId - Order ID (unique transaction reference)
   * @param {number} amount - Payment amount in VND
   * @param {string} orderInfo - Order description
   * @param {string} ipAddr - Customer IP address
   * @param {string} bankCode - Bank code (optional, empty for all banks)
   * @returns {string} VNPay payment URL
   */
  createPaymentUrl(orderId, amount, orderInfo, ipAddr, bankCode = '') {
    const date = new Date();
    const createDate = moment(date).format('YYYYMMDDHHmmss');
    const expireDate = moment(date).add(15, 'minutes').format('YYYYMMDDHHmmss');
    
    let vnp_Params = {
      vnp_Version: '2.1.0',
      vnp_Command: 'pay',
      vnp_TmnCode: this.tmnCode,
      vnp_Locale: 'vn',
      vnp_CurrCode: 'VND',
      vnp_TxnRef: orderId.toString(),
      vnp_OrderInfo: orderInfo,
      vnp_OrderType: 'other',
      vnp_Amount: amount * 100, // VNPay requires amount in smallest unit (VND * 100)
      vnp_ReturnUrl: this.returnUrl,
      vnp_IpAddr: ipAddr,
      vnp_CreateDate: createDate,
      vnp_ExpireDate: expireDate,
    };

    // Add bank code if specified
    if (bankCode) {
      vnp_Params.vnp_BankCode = bankCode;
    }

    // Sort parameters
    vnp_Params = this.sortObject(vnp_Params);
    
    // Create signature
    const signData = querystring.stringify(vnp_Params, { encode: false });
    const hmac = crypto.createHmac('sha512', this.hashSecret);
    const signed = hmac.update(Buffer.from(signData, 'utf-8')).digest('hex');
    
    vnp_Params.vnp_SecureHash = signed;

    // Build final URL
    const paymentUrl = this.vnpUrl + '?' + querystring.stringify(vnp_Params, { encode: false });
    
    return paymentUrl;
  }

  /**
   * Verify VNPay return URL signature
   * 
   * @param {Object} vnp_Params - Query parameters from VNPay return URL
   * @returns {boolean} True if signature is valid
   */
  verifyReturnUrl(vnp_Params) {
    const secureHash = vnp_Params.vnp_SecureHash;
    
    // Remove hash fields before verifying
    const paramsToVerify = { ...vnp_Params };
    delete paramsToVerify.vnp_SecureHash;
    delete paramsToVerify.vnp_SecureHashType;
    
    // Sort parameters
    const sortedParams = this.sortObject(paramsToVerify);
    
    // Create signature
    const signData = querystring.stringify(sortedParams, { encode: false });
    const hmac = crypto.createHmac('sha512', this.hashSecret);
    const signed = hmac.update(Buffer.from(signData, 'utf-8')).digest('hex');

    return secureHash === signed;
  }

  /**
   * Parse VNPay return parameters
   * 
   * @param {Object} vnp_Params - Query parameters from VNPay
   * @returns {Object} Parsed payment result
   */
  parseReturnParams(vnp_Params) {
    return {
      orderId: vnp_Params.vnp_TxnRef,
      amount: parseInt(vnp_Params.vnp_Amount) / 100, // Convert back to VND
      orderInfo: vnp_Params.vnp_OrderInfo,
      responseCode: vnp_Params.vnp_ResponseCode,
      transactionNo: vnp_Params.vnp_TransactionNo,
      bankCode: vnp_Params.vnp_BankCode,
      bankTranNo: vnp_Params.vnp_BankTranNo,
      cardType: vnp_Params.vnp_CardType,
      payDate: vnp_Params.vnp_PayDate,
      transactionStatus: vnp_Params.vnp_TransactionStatus,
    };
  }

  /**
   * Get payment status message from response code
   * 
   * @param {string} responseCode - VNPay response code
   * @returns {string} Status message in Vietnamese
   */
  getResponseMessage(responseCode) {
    const messages = {
      '00': 'Giao dịch thành công',
      '07': 'Trừ tiền thành công. Giao dịch bị nghi ngờ (liên quan tới lừa đảo, giao dịch bất thường)',
      '09': 'Giao dịch không thành công do: Thẻ/Tài khoản của khách hàng chưa đăng ký dịch vụ InternetBanking tại ngân hàng',
      '10': 'Giao dịch không thành công do: Khách hàng xác thực thông tin thẻ/tài khoản không đúng quá 3 lần',
      '11': 'Giao dịch không thành công do: Đã hết hạn chờ thanh toán. Xin quý khách vui lòng thực hiện lại giao dịch',
      '12': 'Giao dịch không thành công do: Thẻ/Tài khoản của khách hàng bị khóa',
      '13': 'Giao dịch không thành công do: Quý khách nhập sai mật khẩu xác thực giao dịch (OTP)',
      '24': 'Giao dịch không thành công do: Khách hàng hủy giao dịch',
      '51': 'Giao dịch không thành công do: Tài khoản của quý khách không đủ số dư để thực hiện giao dịch',
      '65': 'Giao dịch không thành công do: Tài khoản của Quý khách đã vượt quá giới hạn giao dịch trong ngày',
      '75': 'Ngân hàng thanh toán đang bảo trì',
      '79': 'Giao dịch không thành công do: KH nhập sai mật khẩu thanh toán quá số lần quy định',
      '99': 'Các lỗi khác (lỗi còn lại, không có trong danh sách mã lỗi đã liệt kê)',
    };

    return messages[responseCode] || 'Lỗi không xác định';
  }

  /**
   * Check if payment was successful
   * 
   * @param {string} responseCode - VNPay response code
   * @returns {boolean} True if payment successful
   */
  isPaymentSuccess(responseCode) {
    return responseCode === '00';
  }

  /**
   * Create query transaction URL for VNPay API
   * 
   * @param {string} orderId - Order ID
   * @param {string} transDate - Transaction date (YYYYMMDDHHmmss)
   * @param {string} ipAddr - IP address
   * @returns {Object} Query request data
   */
  createQueryRequest(orderId, transDate, ipAddr) {
    const requestData = {
      vnp_RequestId: moment().format('YYYYMMDDHHmmss') + Math.floor(Math.random() * 1000000),
      vnp_Version: '2.1.0',
      vnp_Command: 'querydr',
      vnp_TmnCode: this.tmnCode,
      vnp_TxnRef: orderId.toString(),
      vnp_OrderInfo: `Query transaction ${orderId}`,
      vnp_TransactionDate: transDate,
      vnp_CreateDate: moment().format('YYYYMMDDHHmmss'),
      vnp_IpAddr: ipAddr,
    };

    // Sort and create signature
    const sortedData = this.sortObject(requestData);
    const signData = querystring.stringify(sortedData, { encode: false });
    const hmac = crypto.createHmac('sha512', this.hashSecret);
    const secureHash = hmac.update(Buffer.from(signData, 'utf-8')).digest('hex');
    
    requestData.vnp_SecureHash = secureHash;

    return requestData;
  }

  /**
   * Get list of supported banks
   * 
   * @returns {Array} List of bank objects with code and name
   */
  getSupportedBanks() {
    return [
      { code: '', name: 'Cổng thanh toán VNPAYQR', logo: 'vnpayqr' },
      { code: 'VNPAYQR', name: 'Ứng dụng hỗ trợ VNPAYQR', logo: 'vnpayqr' },
      { code: 'VNBANK', name: 'Thẻ ATM/Tài khoản nội địa', logo: 'vnbank' },
      { code: 'INTCARD', name: 'Thẻ thanh toán quốc tế', logo: 'intcard' },
      { code: 'VIETCOMBANK', name: 'Vietcombank', logo: 'vcb' },
      { code: 'VIETINBANK', name: 'Vietinbank', logo: 'vtb' },
      { code: 'BIDV', name: 'BIDV', logo: 'bidv' },
      { code: 'AGRIBANK', name: 'Agribank', logo: 'agribank' },
      { code: 'SACOMBANK', name: 'Sacombank', logo: 'sacombank' },
      { code: 'TECHCOMBANK', name: 'Techcombank', logo: 'techcombank' },
      { code: 'ACB', name: 'ACB', logo: 'acb' },
      { code: 'VPBANK', name: 'VPBank', logo: 'vpbank' },
      { code: 'TPBANK', name: 'TPBank', logo: 'tpbank' },
      { code: 'MBBANK', name: 'MBBank', logo: 'mbbank' },
      { code: 'SCB', name: 'SCB', logo: 'scb' },
    ];
  }
}

module.exports = new VNPayService();

File: frontend/lib/services/vnpay_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/payment.dart';
import '../config/app_config.dart';

/// VNPay Service for payment integration
///
/// Handles VNPay payment gateway integration including:
/// - Creating payment URLs
/// - Handling payment callbacks
/// - Querying payment status
/// - Managing bank selection
class VNPayService {
  // Singleton pattern
  static final VNPayService _instance = VNPayService._internal();
  factory VNPayService() => _instance;
  VNPayService._internal();

  // Base URL from app config
  final String _baseUrl = AppConfig.baseUrl;

  /// Get authentication token from storage
  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('token');
  }

  /// Create VNPay payment URL
  ///
  /// @param orderId - Order ID to pay for
  /// @param amount - Payment amount in VND
  /// @param orderInfo - Order description (optional)
  /// @param bankCode - Specific bank code (optional, empty for all banks)
  /// @returns PaymentResponse with payment URL
  Future<PaymentResponse> createPayment({
    required int orderId,
    required double amount,
    String? orderInfo,
    String? bankCode,
  }) async {
    try {
      final token = await _getToken();
      if (token == null) {
        throw Exception('Chưa đăng nhập');
      }

      final request = PaymentRequest(
        orderId: orderId,
        amount: amount,
        orderInfo: orderInfo ?? 'Thanh toán đơn hàng #$orderId',
        bankCode: bankCode,
      );

      final response = await http.post(
        Uri.parse('$_baseUrl/payment/vnpay/create'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode(request.toJson()),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return PaymentResponse.fromJson(data);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Lỗi tạo URL thanh toán');
      }
    } catch (error) {
      print('❌ Create payment error: $error');
      rethrow;
    }
  }

  /// Get list of supported banks from VNPay
  ///
  /// @returns List of BankInfo objects
  Future<List<BankInfo>> getSupportedBanks() async {
    try {
      final response = await http.get(
        Uri.parse('$_baseUrl/payment/vnpay/banks'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success'] == true) {
          final List<dynamic> banksJson = data['data'];
          return banksJson.map((json) => BankInfo.fromJson(json)).toList();
        } else {
          throw Exception(data['message'] ?? 'Lỗi lấy danh sách ngân hàng');
        }
      } else {
        throw Exception('Lỗi kết nối server');
      }
    } catch (error) {
      print('❌ Get banks error: $error');
      // Return default bank list if API fails
      return _getDefaultBanks();
    }
  }

  /// Get default bank list (fallback)
  List<BankInfo> _getDefaultBanks() {
    return [
      BankInfo(code: '', name: 'Cổng thanh toán VNPAYQR', logo: 'vnpayqr'),
      BankInfo(
        code: 'VNBANK',
        name: 'Thẻ ATM/Tài khoản nội địa',
        logo: 'vnbank',
      ),
      BankInfo(
        code: 'INTCARD',
        name: 'Thẻ thanh toán quốc tế',
        logo: 'intcard',
      ),
      BankInfo(code: 'VIETCOMBANK', name: 'Vietcombank', logo: 'vcb'),
      BankInfo(code: 'VIETINBANK', name: 'Vietinbank', logo: 'vtb'),
      BankInfo(code: 'BIDV', name: 'BIDV', logo: 'bidv'),
      BankInfo(code: 'AGRIBANK', name: 'Agribank', logo: 'agribank'),
      BankInfo(code: 'TECHCOMBANK', name: 'Techcombank', logo: 'techcombank'),
      BankInfo(code: 'ACB', name: 'ACB', logo: 'acb'),
      BankInfo(code: 'VPBANK', name: 'VPBank', logo: 'vpbank'),
      BankInfo(code: 'MBBANK', name: 'MBBank', logo: 'mbbank'),
    ];
  }

  /// Handle VNPay payment callback from WebView
  ///
  /// @param params - Query parameters from VNPay return URL
  /// @returns PaymentResult with payment status
  Future<PaymentResult> handlePaymentCallback(
    Map<String, String> params,
  ) async {
    try {
      final response = await http.post(
        Uri.parse('$_baseUrl/payment/vnpay/callback'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(params),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return PaymentResult.fromJson(data);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Lỗi xử lý callback thanh toán');
      }
    } catch (error) {
      print('❌ Handle callback error: $error');
      rethrow;
    }
  }

  /// Parse VNPay return URL parameters
  ///
  /// @param url - Return URL from VNPay
  /// @returns Map of query parameters
  Map<String, String> parseReturnUrl(String url) {
    try {
      final uri = Uri.parse(url);
      return uri.queryParameters;
    } catch (error) {
      print('❌ Parse URL error: $error');
      return {};
    }
  }

  /// Get payment status for an order
  ///
  /// @param orderId - Order ID to check status
  /// @returns PaymentStatus object
  Future<PaymentStatus> getPaymentStatus(int orderId) async {
    try {
      final token = await _getToken();
      if (token == null) {
        throw Exception('Chưa đăng nhập');
      }

      final response = await http.get(
        Uri.parse('$_baseUrl/payment/vnpay/status/$orderId'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        if (data['success'] == true) {
          return PaymentStatus.fromJson(data['data']);
        } else {
          throw Exception(data['message'] ?? 'Lỗi lấy trạng thái thanh toán');
        }
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Lỗi kết nối server');
      }
    } catch (error) {
      print('❌ Get payment status error: $error');
      rethrow;
    }
  }

  /// Check if URL is VNPay return URL
  ///
  /// @param url - URL to check
  /// @returns true if URL is VNPay return URL
  bool isVNPayReturnUrl(String url) {
    return url.contains('vnp_ResponseCode') &&
        url.contains('vnp_TransactionNo') &&
        url.contains('vnp_TxnRef');
  }

  /// Extract order number from VNPay return URL
  ///
  /// @param url - VNPay return URL
  /// @returns Order number (vnp_TxnRef)
  String? extractOrderNumber(String url) {
    try {
      final uri = Uri.parse(url);
      return uri.queryParameters['vnp_TxnRef'];
    } catch (error) {
      print('❌ Extract order number error: $error');
      return null;
    }
  }

  /// Check if payment was successful from URL
  ///
  /// @param url - VNPay return URL
  /// @returns true if payment successful (responseCode = 00)
  bool isPaymentSuccessFromUrl(String url) {
    try {
      final uri = Uri.parse(url);
      final responseCode = uri.queryParameters['vnp_ResponseCode'];
      return responseCode == '00';
    } catch (error) {
      print('❌ Check payment success error: $error');
      return false;
    }
  }

  /// Get response message from VNPay response code
  ///
  /// @param responseCode - VNPay response code
  /// @returns Vietnamese message
  String getResponseMessage(String responseCode) {
    final messages = {
      '00': 'Giao dịch thành công',
      '07': 'Trừ tiền thành công. Giao dịch bị nghi ngờ',
      '09': 'Thẻ/Tài khoản chưa đăng ký InternetBanking',
      '10': 'Xác thực thông tin không đúng quá 3 lần',
      '11': 'Đã hết hạn chờ thanh toán',
      '12': 'Thẻ/Tài khoản bị khóa',
      '13': 'Nhập sai mật khẩu OTP',
      '24': 'Khách hàng hủy giao dịch',
      '51': 'Tài khoản không đủ số dư',
      '65': 'Vượt quá giới hạn giao dịch trong ngày',
      '75': 'Ngân hàng thanh toán đang bảo trì',
      '79': 'Nhập sai mật khẩu quá số lần quy định',
      '99': 'Lỗi không xác định',
    };

    return messages[responseCode] ?? 'Lỗi không xác định';
  }
}
